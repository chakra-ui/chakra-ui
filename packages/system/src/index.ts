export * from "@chakra-ui/color-mode"
export * from "@chakra-ui/styled-system"
export { keyframes } from "@emotion/react"
export type { Interpolation } from "@emotion/react"
export * from "./hooks"
export * from "./system.types"
export * from "./providers"
export { omitThemingProps } from "./system.utils"
export * from "./system"
export * from "./forward-ref"
export * from "./use-style-config"
export { shouldForwardProp } from "./should-forward-prop"
