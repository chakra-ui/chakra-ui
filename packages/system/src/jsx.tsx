import { Dict, runIfFn } from "@chakra-ui/utils"
import { jsx as emotion } from "@emotion/core"
import { SystemStyleObject, css } from "@chakra-ui/styled-system"

interface GetCSS {
  sx?: any
  css?: any
}

function getCSS(props: GetCSS) {
  if (!props.sx && !props.css) return undefined
  /**
   * Leverage emotion's css function interpolation to access the theme
   */
  return (theme: Dict) => {
    /**
     * process the theme-aware cx prop
     */
    const sxStyles = css(props.sx)(theme)
    /**
     * process the normal emotion's css prop
     * (NB: This is not theme-aware, and you can't use shorthand style props)
     */
    const cssStyles = runIfFn(props.css, theme)
    /**
     * return an array value and allow emotion do the rest.
     * By default, emotion can handle array style values
     */
    return [cssStyles, sxStyles]
  }
}

function parse(props: Dict | undefined) {
  if (!props) return null

  const computedProps: Dict = {}

  for (const prop in props) {
    if (prop === "sx") continue
    computedProps[prop] = props[prop]
  }

  const css = getCSS(props)

  if (css) computedProps.css = css

  return computedProps
}

export const jsx = (
  type: React.ElementType,
  props: Dict,
  ...children: React.ReactNode[]
) => emotion.apply(undefined, [type, parse(props), ...children])

interface SxProp {
  sx?: SystemStyleObject
}

declare module "react" {
  interface Attributes extends SxProp {}
}

declare global {
  // eslint-disable-next-line
  namespace JSX {
    interface IntrinsicAttributes extends SxProp {}
  }
}

export default jsx
