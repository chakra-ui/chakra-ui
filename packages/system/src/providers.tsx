import { useColorMode } from "@chakra-ui/color-mode"
import {
  css,
  SystemStyleObject,
  toCSSVar,
  WithCSSVar,
} from "@chakra-ui/styled-system"
import { Dict, memoizedGet as get, runIfFn } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import {
  Global,
  Interpolation,
  ThemeContext,
  ThemeProvider as EmotionThemeProvider,
  ThemeProviderProps as EmotionThemeProviderProps,
} from "@emotion/react"
import * as React from "react"

export interface ThemeProviderProps extends EmotionThemeProviderProps {
  cssVarsRoot?: string
}

export const ThemeProvider = (props: ThemeProviderProps) => {
  const { cssVarsRoot, theme, children } = props
  const computedTheme = React.useMemo(() => toCSSVar(theme), [theme])
  return (
    <EmotionThemeProvider theme={computedTheme}>
      <CSSVars root={cssVarsRoot} />
      {children}
    </EmotionThemeProvider>
  )
}

export interface CSSVarsProps {
  /**
   * The element to attach the CSS custom properties to.
   * @default ":host, :root"
   */
  root?: string
}

export const CSSVars = ({ root = ":host, :root" }: CSSVarsProps) => {
  /**
   * Append color mode selector to allow semantic tokens to change according to the color mode
   */
  const selector = [root, `[data-theme]`].join(",")
  return <Global styles={(theme: any) => ({ [selector]: theme.__cssVars })} />
}

export function useTheme<T extends object = Dict>() {
  const theme = React.useContext(
    ThemeContext as unknown as React.Context<T | undefined>,
  )
  if (!theme) {
    throw Error(
      "useTheme: `theme` is undefined. Seems you forgot to wrap your app in `<ChakraProvider />` or `<ThemeProvider />`",
    )
  }

  return theme as WithCSSVar<T>
}

const [StylesProvider, useStyles] = createContext<Dict<SystemStyleObject>>({
  name: "StylesContext",
  errorMessage:
    "useStyles: `styles` is undefined. Seems you forgot to wrap the components in `<StylesProvider />` ",
})
export { StylesProvider, useStyles }

/**
 * Applies styles defined in `theme.styles.global` globally
 * using emotion's `Global` component
 */
export const GlobalStyle = () => {
  const { colorMode } = useColorMode()
  return (
    <Global
      styles={(theme: any) => {
        const styleObjectOrFn = get(theme, "styles.global")
        const globalStyles = runIfFn(styleObjectOrFn, { theme, colorMode })
        if (!globalStyles) return undefined
        const styles = css(globalStyles)(theme)
        return styles as Interpolation<{}>
      }}
    />
  )
}
