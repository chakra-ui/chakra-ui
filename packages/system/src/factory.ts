import { DOMElements } from "./system.utils"
import { ChakraStyledOptions, HTMLChakraComponents, styled } from "./system"
import { As, ChakraComponent } from "./system.types"

type ChakraFactory = {
  <T extends As, P = {}>(
    component: T,
    options?: ChakraStyledOptions,
  ): ChakraComponent<T, P>
}

function factory() {
  const cache = new Map<DOMElements, ChakraComponent<DOMElements>>()

  return new Proxy(styled, {
    /**
     * @example
     * const Div = chakra("div")
     * const WithChakra = chakra(AnotherComponent)
     */
    apply(target, thisArg, argArray: [DOMElements, ChakraStyledOptions]) {
      return styled(...argArray)
    },
    /**
     * @example
     * <chakra.div />
     */
    get(_, element: DOMElements) {
      if (!cache.has(element)) {
        cache.set(element, styled(element))
      }
      return cache.get(element)
    },
  }) as ChakraFactory & HTMLChakraComponents
}

export const chakra = factory()
