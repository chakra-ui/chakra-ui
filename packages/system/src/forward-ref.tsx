/**
 * All credit goes to Chance (Reach UI), Haz (Reakit) and (fluentui)
 * for creating the base type definitions upon which we improved on
 */
import * as React from "react"
import { As, ComponentWithAs, PropsOf, RightJoinProps } from "./system.types"

export function forwardRef<Props extends object, Component extends As>(
  component: React.ForwardRefRenderFunction<
    any,
    RightJoinProps<PropsOf<Component>, Props> & {
      as?: As
    }
  >,
) {
  return (React.forwardRef(component) as unknown) as ComponentWithAs<
    Component,
    Props
  >
}
