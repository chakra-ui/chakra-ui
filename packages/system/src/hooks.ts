import { useColorMode } from "@chakra-ui/color-mode"
import { SystemStyleObject } from "@chakra-ui/styled-system"
import {
  Dict,
  filterUndefined,
  mergeWith,
  runIfFn,
  StringOrNumber,
} from "@chakra-ui/utils"
import { useMemo, useRef } from "react"
import isEqual from "react-fast-compare"
import { useTheme } from "./providers"
import { ThemingProps } from "./system.types"
import { omitThemingProps } from "./system.utils"

export function useChakra<T extends Dict = Dict>() {
  const colorModeResult = useColorMode()
  const theme = useTheme() as T
  return { ...colorModeResult, theme }
}

const resolveBreakpointValue = <T extends StringOrNumber>(
  theme: Dict,
  tokenValue: T,
  fallbackValue: any,
) => {
  if (tokenValue === null) return tokenValue
  const getValue = (val: T) => theme.__breakpoints?.asArray?.[val]
  return getValue(tokenValue) ?? getValue(fallbackValue) ?? fallbackValue
}

// inspired from ./css.ts : resolveTokenValue
const resolveTokenValue = <T extends StringOrNumber>(
  theme: Dict,
  tokenValue: T,
  fallbackValue: any,
) => {
  if (tokenValue == null) return tokenValue
  const getValue = (val: T) => theme.__cssMap?.[val]?.value
  return getValue(tokenValue) ?? getValue(fallbackValue) ?? fallbackValue
}

export function useToken<T extends StringOrNumber>(
  scale: string,
  token: T | T[],
  fallback?: T | T[],
) {
  const theme = useTheme()

  if (Array.isArray(token)) {
    let fallbackArr: T[] = []
    if (fallback) {
      fallbackArr = Array.isArray(fallback) ? fallback : [fallback]
    }

    return token.map((token, index) => {
      if (scale === "breakpoints") {
        return resolveBreakpointValue(theme, token, fallbackArr[index] ?? token)
      }
      const path = `${scale}.${token}`
      return resolveTokenValue(theme, path, fallbackArr[index] ?? token)
    })
  }

  if (scale === "breakpoints") {
    return resolveBreakpointValue(theme, token, fallback)
  }

  const path = `${scale}.${token}`
  return resolveTokenValue(theme, path, fallback)
}

export function useProps<P extends ThemingProps>(
  themeKey: string,
  props: P,
  isMulti: true,
): {
  styles: Record<string, SystemStyleObject>
  props: Omit<P, keyof ThemingProps>
}

export function useProps<P extends ThemingProps>(
  themeKey: string,
  props?: P,
  isMulti?: boolean,
): {
  styles: SystemStyleObject
  props: Omit<P, keyof ThemingProps>
}

export function useProps(themeKey: string, props: Dict) {
  const { theme, colorMode } = useChakra()

  const styleConfig = (props.styleConfig || theme.components?.[themeKey]) as
    | Dict
    | undefined

  const defaultProps = styleConfig?.defaultProps ?? {}
  const propsWithDefault = { ...defaultProps, ...filterUndefined(props) }

  const stylesRef = useRef<Dict>({})

  const mergedProps = mergeWith({}, propsWithDefault, { theme, colorMode })

  const memoizedStyles = useMemo(() => {
    if (styleConfig) {
      const baseStyles = runIfFn(styleConfig.baseStyle ?? {}, mergedProps)

      const variants = runIfFn(
        styleConfig.variants?.[mergedProps.variant as string] ?? {},
        mergedProps,
      )

      const sizes = runIfFn(
        styleConfig.sizes?.[mergedProps.size as string] ?? {},
        mergedProps,
      )

      const styles = mergeWith(baseStyles, sizes, variants)

      if (styleConfig.parts) {
        styleConfig.parts.forEach((part: string) => {
          styles[part] = styles[part] ?? {}
        })
      }

      const isStyleEqual = isEqual(stylesRef.current, styles)

      if (!isStyleEqual) {
        stylesRef.current = styles
      }
    }

    return stylesRef.current
  }, [styleConfig, mergedProps])

  return {
    styles: memoizedStyles,
    props: omitThemingProps(propsWithDefault),
  }
}
