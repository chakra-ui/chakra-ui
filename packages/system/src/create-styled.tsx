import { css, getComponentStyles } from "@chakra-ui/parser"
import {
  As,
  Dict,
  isEmptyObject,
  isString,
  runIfFn,
  isUndefined,
} from "@chakra-ui/utils"
import { CSSObject } from "@emotion/core"
import hoist from "hoist-non-react-statics"
import * as React from "react"
import { useChakra } from "./hooks"
import jsx from "./jsx"
import {
  customShouldForwardProp,
  filterProps,
  removeStyleProps,
} from "./should-forward-prop"
import { ChakraComponent, Options } from "./styled.types"
import { getDisplayName } from "./styled.utils"

function createStyled<T extends As, P>(component: T, options?: Options<T, P>) {
  return function(...interpolations: any[]) {
    const Styled = React.forwardRef(
      ({ as, ...props }: any, ref: React.Ref<any>) => {
        // Get the color mode and theme from context
        const { theme, colorMode } = useChakra()

        // We'll store the final css object of style props here
        let computedStyles: CSSObject = {}

        // For each style interpolation, we'll pass the theme and colorMode
        const propsWithTheme = { theme, colorMode, ...props }

        /**
         * Users can pass a base style to the component options, let's resolve it
         *
         * @example
         * const Button = chakra("button", {
         *  baseStyle: {
         *    margin: 4,
         *    color: "red.300"
         *  }
         * })
         */
        if (options?.baseStyle) {
          const baseStyleObject = runIfFn(options.baseStyle, propsWithTheme)
          const baseStyle = css(baseStyleObject)(theme)
          computedStyles = { ...computedStyles, ...baseStyle }
        }

        /**
         * Users can pass a theme key to reference styles in the theme, let's resolve it.
         * Styles will be read from `theme.components.<themeKey>`
         *
         * @example
         * const Button = chakra("button", {
         *  themeKey: "Button"
         * })
         */
        if (options?.themeKey) {
          const componentStyles = getComponentStyles(
            propsWithTheme,
            options.themeKey as string,
          )
          computedStyles = { ...computedStyles, ...componentStyles }
        }

        // Resolve each interpolation and add result to final style
        interpolations.forEach(interpolation => {
          const style = runIfFn(interpolation, propsWithTheme)
          computedStyles = { ...computedStyles, ...style }
        })

        const element = as || component
        const isTag = isString(element)

        let computedProps: Dict = isTag ? filterProps(props) : { ...props }

        /**
         * Users can pass a html attributes to component options, let's resolve it.
         * Attributes will be passed to the underlying dom element
         *
         * @example
         * const Button = chakra("button", {
         *  attrs: {
         *    type: "submit"
         *  }
         * })
         */
        if (options?.attrs) {
          const attrsProps = runIfFn(options.attrs, propsWithTheme)
          computedProps = { ...computedProps, ...attrsProps }
        }

        if (!isTag) computedProps = removeStyleProps(computedProps)

        /**
         * Users can pass an option to control how props are forwarded
         *
         * @example
         * const Button = chakra("button", {
         *  attrs: props => ({
         *    type: "submit",
         *    disabled: props.isDisabled
         *  }),
         * shouldForwardProps: prop => prop !== "isDisabled"
         * })
         */
        if (options?.shouldForwardProp) {
          computedProps = customShouldForwardProp(
            options.shouldForwardProp,
            computedProps,
          )
        }

        // check if style is empty, we don't want to pass css prop to jsx if it's empty
        const isStyleEmpty = isEmptyObject(computedStyles)

        computedProps.css = isStyleEmpty
          ? runIfFn(computedProps.css, theme)
          : { ...computedStyles, ...runIfFn(computedProps.css, theme) }

        /**
         * This helps to prevent scenarios where no styles was passed
         * to the component but emotion generate a `css-0` className.
         */
        if (
          isEmptyObject(computedProps.css) ||
          isUndefined(computedProps.css)
        ) {
          delete computedProps.css
        }

        /**
         * Create the element using emotion's jsx, similar to React.createElement
         * but it allows us pass a css object as prop and it'll convert it to a className
         */
        return jsx(element, { ref, ...computedProps })
      },
    )

    // Compute the display name of the final component
    Styled.displayName = getDisplayName(component)

    // [Optimization] users can pass a pure option to memoize this component
    const StyledComponent = options?.pure ? React.memo(Styled) : Styled

    // hoist all non-react statics attached to the `component` prop
    const Component = hoist(
      StyledComponent,
      component as React.ComponentType<any>,
    )

    return Component as ChakraComponent<T, P>
  }
}

export default createStyled
