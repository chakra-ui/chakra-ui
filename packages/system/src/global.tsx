import { useColorMode } from "@chakra-ui/color-mode/src"
import css from "@chakra-ui/css/src"
import { get, runIfFn } from "@chakra-ui/utils/src"
import { Global, Interpolation } from "@emotion/core"
import * as React from "react"

export function GlobalStyle() {
  const { colorMode } = useColorMode()
  return (
    <Global
      styles={(theme) => {
        const styleObjectOrFn = get(theme, "styles.global")
        const styleObject = runIfFn(styleObjectOrFn, { theme, colorMode })
        if (!styleObject) return
        const styles = css({ body: styleObject })(theme)
        return styles as Interpolation
      }}
    />
  )
}
