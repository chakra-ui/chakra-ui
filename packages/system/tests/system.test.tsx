import * as React from "react"
import renderer from "react-test-renderer"
import { chakra } from "../src"

const renderJSON = (el: React.ReactElement) => renderer.create(el).toJSON()

test("should render correctly", () => {
  const json = renderJSON(<chakra.div mt="40px">welcome</chakra.div>)
  expect(json).toMatchSnapshot()
})
