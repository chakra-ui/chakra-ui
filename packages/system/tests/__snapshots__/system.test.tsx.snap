// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="css-y1gt6f"
>
  welcome
</div>
`;
